examples/communication/grandchild_to_grandparent/src/main.rs

```diff
@@ -0,0 +1,145 @@
+use std::rc::Rc;
+use yew::{html, Callback, Component, Context, ContextHandle, ContextProvider, Html};
+
+// This is the shared state between the parent and child components.
+#[derive(Clone, PartialEq)]
+pub struct AppState {
+    value: u32,
+    child_clicked: Callback<()>,
+}
+
+// Our top-level component.
+
+pub struct GrandParent {
+    state: Rc<AppState>,
+}
+
+pub enum Msg {
+    ButtonClick,
+}
+
+impl Component for GrandParent {
+    type Message = Msg;
+    type Properties = ();
+
+    fn create(ctx: &Context<Self>) -> Self {
+        let child_clicked = ctx.link().callback(|_| Msg::ButtonClick);
+        let state = Rc::new(AppState {
+            value: 0,
+            child_clicked,
+        });
+        Self { state }
+    }
+
+    fn update(&mut self, _ctx: &Context<Self>, msg: Self::Message) -> bool {
+        match msg {
+            Msg::ButtonClick => {
+                let value = self.state.value + 1;
+                self.state = Rc::new(AppState {
+                    value,
+                    ..(*self.state).clone()
+                });
+                true
+            }
+        }
+    }
+
+    fn view(&self, _ctx: &Context<Self>) -> Html {
+        let app_state = self.state.clone();
+        let msg = format!(
+            "My grandchildren have been clicked {} times",
+            self.state.value
+        );
+
+        html! {
+            <ContextProvider<Rc<AppState>> context={app_state}>
+                <div class="app">
+                    <div class="parent">
+                        <h2>{ "Grandchild-to-Grandparent (and back!) communication example" }</h2>
+                        <div>{msg}</div>
+                        <Parent />
+                    </div>
+                </div>
+            </ContextProvider<Rc<AppState>>>
+        }
+    }
+}
+
+// The `Parent` component is the parent of the `Child` component. It has no logic, and is here to show there is no direct
+// relation between grandchild and grandparent.
+
+pub struct Parent;
+
+impl Component for Parent {
+    type Message = ();
+    type Properties = ();
+
+    fn create(_ctx: &Context<Self>) -> Self {
+        Self
+    }
+
+    fn view(&self, _ctx: &Context<Self>) -> Html {
+        html! {
+            <>
+                <Child />
+                <Child />
+            </>
+        }
+    }
+}
+
+// The `Child` component is the child of the `Parent` component, and will send and receive updates to/from the grandparent using the context.
+
+pub struct Child {
+    state: Rc<AppState>,
+    _listener: ContextHandle<Rc<AppState>>,
+}
+
+pub enum ChildMsg {
+    ContextChanged(Rc<AppState>),
+}
+
+impl Component for Child {
+    type Message = ChildMsg;
+    type Properties = ();
+
+    fn create(ctx: &Context<Self>) -> Self {
+        let (state, _listener) = ctx
+            .link()
+            .context::<Rc<AppState>>(ctx.link().callback(ChildMsg::ContextChanged))
+            .expect("context to be set");
```

Comment:
I think it would be good to have a `FunctionChild` that shows the usage of [`use_context`](https://yew-rs-api.web.app/next/yew/functional/fn.use_context.html) for comparison and completeness.

Reply:
I get your point, but keeping the communication examples structurally the same helps people spot the differences between them. 

The use of `use_context` in a functional component is demonstrated by the `examples/contexts` example. Maybe we can include a comment here to point this out explicitly: "See the `context` example for a demonstration of the use of contexts in a functional component." ?

Reply:
Yeah a comment sounds good, I agree that a single style is easier to read.